### PR TITLE
fix: batch metric writes to TensorBoard [MLG-990]

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description

This fixes an issue where writing metrics to TensorBoard less than 1 second apart will cause the previous file to be overwritten. This is due to TensorFlow's `tfevent` files being named by the epoch second of their creation. We can't control this filename and/or override the epoch second, since this is implemented by TensorFlow in the C code.

Previously, we "reset" (close the current event file and open a new one) the writer every time a metric is written. The approach here is to only reset the file after at least 1 second has elapsed since the last write, effectively batching the events so that a new file is written at maximum once per second and each file might contain more than one event.

This isn't a _perfect_ solution, because we keep track of the last write time on our end, but it's not 100% guaranteed to be the ultimate epoch time TensorFlow names the file with. For instance, if the timestamp we get is 0.999999 epoch seconds, but by the time the TF writer persists the `tfevent` file it is named 1, we would still treat the next write as 1 second and overwrite the file. 

I'm not sure how to measure the likelihood this would happen, but it seems relatively low, and this fix is an improvement over not having batching at all. We could check for the latest `tfevent` file written in the `log_dir` directory and go off that timestamp instead, but this introduces some overhead into writing since TF doesn't surface a way to get the current file so we'd have to iterate through every file in the `log_dir` directory.

## Test Plan

Run an experiment that reports metrics more than once per second. Verify that all the metrics reported exist in TensorBoard and/or `log_dir` `tfevent` files.

With Core API:

```
import logging

import determined as det


def main(core_context):
    for batch in range(100):
        step = batch + 1
        core_context.train.report_training_metrics(
            steps_completed=step, metrics={"x": step}
        )
        if step % 10 == 0:
            core_context.train.report_validation_metrics(
                steps_completed=step,
                metrics={"x": step}
            )


if __name__ == "__main__":
    logging.basicConfig(level=logging.DEBUG, format=det.LOG_FORMAT)

    with det.core.init() as core_context:
        main(core_context=core_context)
```
```
name: metrics
entrypoint: python3 metrics.py

searcher:
   name: single
   metric: x
   max_length: 1

max_restarts: 0
```

TensorBoard should show 100 training steps and 10 validation steps. There should be a smaller fraction of `tfevent` files (depending on how long it takes to run the script, 6 seconds = 6 event files on my local machine) present in whatever `log_dir` directory is configured.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

**Review by rb-determined-ai — COMMENTED** [review]
can you include PyTorchTrial in this fix?

Instead of using the deprecated TorchWriter, it configures core.TensorBoard.MANUAL_MODE and uses a standard pytorch SummaryWriter.

File: harness/tests/tensorboard/test_metric_writers.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        if i % validation_period == 0:
+            batch_writer.on_validation_step_end(steps_completed=i, metrics={"x": step})
+
+    train_events = []
+    val_events = []
+
+    # Read event files saved and verify all metrics are written.
+    event_files = list(tmp_path.iterdir())
+    for file in event_files:
+        print(file)
```
Comment (ioga) [inline harness/tests/tensorboard/test_metric_writers.py:63]:
```suggestion
```

maybe remove this debug print? pytest will hide these by default and if someone will be debugging a failure here, they'll probably want their own prints.

File: harness/tests/tensorboard/test_metric_writers.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        for event in summary_iterator.summary_iterator(str(file)):
+            # TensorFlow injects an event containing metadata at the start of every tfevent
+            # file; ignore these.
+            if getattr(event, "file_version", None):
+                continue
+            for event_data in event.summary.value:
+                print(event_data)
```
Comment (ioga) [inline harness/tests/tensorboard/test_metric_writers.py:70]:
```suggestion
```

**Review by ioga — APPROVED** [review]
lgtm

File: harness/determined/pytorch/_pytorch_context.py [hunk 3]
```diff
@@ -1037,8 +1040,22 @@ def train_batch(self, batch, epoch_idx, batch_idx):
         return self._tbd_writer
 
     def _maybe_reset_tbd_writer(self) -> None:
-        if self._tbd_writer is not None:
+        """
+        Reset the TensorBoard writer (close so the next write will reopen a new file) if more than
+        1 second has elapsed since the last reset. This is to ensure writes < 1 second apart do
+        not overwrite the same file name generated by TensorBoard from the epoch second.
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_context.py:1046]: Recommend rephrasing this to be more literally what you do in the code.

Or make the code more literally what you do in the docstring.

Because what the code is doing now is _not_ checking that subsequent resets occurs at least a second later; it is checking that the next reset occurs within a different integer second.

I don't care which you do but docstrings that offer close-but-not-correct explanations are of very limited use.

File: harness/determined/pytorch/_pytorch_context.py [hunk 4]
```diff
@@ -133,6 +134,8 @@ def __init__(
 
         self._tbd_writer = None  # type: Optional[Any]
         self._enable_tensorboard_logging = enable_tensorboard_logging
+        # Timestamp for batching TensorBoard uploads
+        self._last_tb_reset_ts: Optional[float] = None
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_context.py:138]: I recommend avoiding `Optional` and defaulting to `0.0` instead; it would simplify the code.  As long as nobody runs this code before 1970, I suppose.

File: harness/determined/tensorboard/metric_writers/callback.py [hunk 5]
```diff
@@ -19,10 +20,15 @@ def add_scalar(self, name: str, value: Union[int, float, "np.number"], step: int
...
 
+    @abc.abstractmethod
+    def flush(self) -> None:
+        pass
+
 
 class BatchMetricWriter:
     def __init__(self, writer: MetricWriter) -> None:
         self.writer = writer
+        self._last_flush_ts: Optional[float] = None
```
Comment (rb-determined-ai) [inline harness/determined/tensorboard/metric_writers/callback.py:31]: ditto on the timing.

File: harness/determined/tensorboard/metric_writers/callback.py [hunk 6]
```diff
@@ -54,13 +62,36 @@ def on_train_step_end(
...
+        other.
+
+        This effectively batches event writes so each event file may contain more than one event.
+        """
+        current_ts = time.time()
+        if not self._last_flush_ts:
+            return
+
+        if int(current_ts) > int(self._last_flush_ts):
+            self.writer.reset()
```
Comment (rb-determined-ai) [inline harness/determined/tensorboard/metric_writers/callback.py:97]: why the different implementation and different docstring?

In `_pytorch_context.py`, the reset logic was entirely encapsulated in `_maybe_reset_tbd_writer()`, but here it's spread out?  Why is that?

File: harness/determined/pytorch/_pytorch_context.py (hunk 4 above)
Comment (azhou-determined) [reply harness/determined/pytorch/_pytorch_context.py:138]: this would cause the first event file to be overwritten. the initial reset ts needs to be a value that _doesn't_ cause the reset logic to trigger.

otherwise what happens is we would reset the initial file then open a new one (usually within the same second), and the original file would get overwritten.

File: harness/determined/tensorboard/metric_writers/callback.py (hunk 6 above)
Comment (azhou-determined) [reply harness/determined/tensorboard/metric_writers/callback.py]: discussed offline, updated

File: harness/determined/tensorboard/metric_writers/callback.py [hunk 7]
```diff
@@ -54,13 +60,35 @@ def on_train_step_end(
...
+        Reset (close current file and open a new one) the current writer if the current epoch
+        second is at least one second greater than the epoch second of the last reset.
+
+        The TensorFlow event writer names each event file by the epoch second it is created, so
+        if events are written quickly in succession (< 1 second apart), they will overwrite each
+        other.
+
+        This effectively batches event writes so each event file may contain more than one event.
+        """
+        self.writer.flush()
```
Comment (rb-determined-ai) [inline harness/determined/tensorboard/metric_writers/callback.py:86]: Wait, why flush in cases where we are not going to reset?
